File: 0301-remove-invalid-parentheses/0301-remove-invalid-parentheses.java
class Solution {
    private void backtrack(int len, int open, int close, int minLength, String curr, String s,
                                int ind, Set<String> list) {
        if(close > open) return;
        if(ind >= len) {
            if(open == close) {
                if(curr.length() == minLength) {
                    // System.out.println("Add: " + curr);
                    list.add(curr);
                }
            }
            // System.out.println("skip: " + curr);
            return;
        }
        if(s.charAt(ind) == '(') {
            String newStr = new StringBuilder(curr).append('(').toString();
            backtrack(len, open+1, close, minLength, newStr, s, ind+1, list);
            backtrack(len, open, close, minLength, curr, s, ind+1, list);
        } else if (s.charAt(ind) == ')') {
            String newStr = new StringBuilder(curr).append(')').toString();
            backtrack(len, open, close+1, minLength, newStr, s, ind+1, list);
            backtrack(len, open, close, minLength, curr, s, ind+1, list);
        } else {
            String newStr = new StringBuilder(curr).append(s.charAt(ind)).toString();
            backtrack(len, open, close, minLength, newStr, s, ind+1, list);
        }
    }
    public List<String> removeInvalidParentheses(String s) {
        int minToRemove = 0;
        int opening = 0;
        int closing = 0;
        int len = s.length();
        Stack<Character> stk = new Stack();
        
        for(int i=0; i<len; i++) {
            char ch = s.charAt(i);
            if(ch != '(' && ch !=')') continue;
            if(stk.size() == 0) {
                stk.add(ch);
            } else {
                char peek = stk.peek();
                if(peek == '(' && ch == ')') stk.pop();
                else {
                    stk.add(ch);
                }
            }
        }
        // System.out.println("min remove: " + stk.size());
        minToRemove = stk.size();
        Set<String> res = new HashSet();
        backtrack(len, 0, 0, len-minToRemove, "", s, 0, res);
        if(res.size() == 0) {
            res.add("");
        }
        return new ArrayList(res);
    }
}

File: 0301-remove-invalid-parentheses/0301-remove-invalid-parentheses.js
/**
 * @param {string} s
 * @return {string[]}
 */
function backtrack(s, curr, ind, open, close, res) {
    if(ind >= s.length) {
        if(open === close && curr != "") {
            res.add(curr)
        }
        return
    }

    if(close > open) {
        return
    }
    const ch = s.charAt(ind)
    if(ch === '(') {
        backtrack(s, `${curr}(`, ind+1, open+1, close, res)
    } else if(ch === ')') {
        backtrack(s, `${curr})`, ind+1, open, close+1, res)
    } else {
        backtrack(s, `${curr}${ch}`, ind+1, open, close, res)
        return
    }
    backtrack(s, `${curr}`, ind+1, open, close, res)
}
var removeInvalidParentheses = function(s) {
    const set = new Set()
    backtrack(s, '', 0, 0, 0, set)
    if(set.size === 0) return [""]
    const arr = [...set.values()]
    arr.sort((a, b) => b.length - a.length)
    const res = []
    const maxlen = arr[0].length
    for(const str of arr) {
        if(str.length === maxlen){
            res.push(str)
        }
    }
    // console.log(arr)
    return res
};